src/constants.rs

```diff
@@ -1,4 +1,333 @@
 //! Constants used in the Orchard protocol.
+use ff::{Field, PrimeField};
+use group::Curve;
+use halo2::{
+    arithmetic::{lagrange_interpolate, CurveAffine, FieldExt},
+    pasta::pallas,
+};
+
+pub mod commit_ivk_r;
+pub mod note_commit_r;
+pub mod nullifier_k;
+pub mod value_commit_r;
+pub mod value_commit_v;
+
+pub mod util;
 
 /// $\ell^\mathsf{Orchard}_\mathsf{base}$
 pub(crate) const L_ORCHARD_BASE: usize = 255;
+
+/// $\ell_\mathsf{value}$
+pub(crate) const L_VALUE: usize = 64;
+
+// SWU hash-to-curve personalizations
+
+/// SWU hash-to-curve personalization
+/// This is used for the spending key base point and the nullifier base point K^Orchard
+pub const ORCHARD_PERSONALIZATION: &str = "z.cash:Orchard";
+
+/// SWU hash-to-curve personalization for the group hash for key diversification
+pub const KEY_DIVERSIFICATION_PERSONALIZATION: &str = "z.cash:Orchard-gd";
+
+/// SWU hash-to-curve personalization for the value commitment generator
+pub const VALUE_COMMITMENT_PERSONALIZATION: &str = "z.cash:Orchard-cv";
+
+/// SWU hash-to-curve personalization for the note commitment generator
+pub const NOTE_COMMITMENT_PERSONALIZATION: &str = "z.cash:Orchard-NoteCommit";
+
+/// SWU hash-to-curve personalization for the IVK commitment generator
+pub const COMMIT_IVK_PERSONALIZATION: &str = "z.cash:Orchard-CommitIvk";
+
+/// SWU hash-to-curve personalization for the Merkle CRH generator
+pub const MERKLE_CRH_PERSONALIZATION: &str = "z.cash:Orchard-MerkleCRH";
+
+/// Window size for fixed-base scalar multiplication
+pub const FIXED_BASE_WINDOW_SIZE: usize = 3;
+
+/// $2^{`FIXED_BASE_WINDOW_SIZE`}$
+pub const H: usize = 1 << FIXED_BASE_WINDOW_SIZE;
+
+/// Number of windows for a full-width scalar
+pub const NUM_WINDOWS: usize = pallas::Base::NUM_BITS as usize / FIXED_BASE_WINDOW_SIZE;
```

Comment:
This is rounded down; is that correct?

Reply:
Oh, 255 is a multiple of 3. But that's dependent on the specific values.

Reply:
:+1: I will change this to round it up.